sumcheck/src/verifier/mod.rs

```diff
@@ -86,12 +86,14 @@ pub fn verify_sum_check_input_layer<E: FieldElement, H: ElementHasher<BaseField
 
     let mut numerators = vec![E::ZERO; evaluator.get_num_fractions()];
     let mut denominators = vec![E::ZERO; evaluator.get_num_fractions()];
-    evaluator.evaluate_query(
-        &proof.0.openings_claim.openings,
-        &log_up_randomness,
-        &mut numerators,
-        &mut denominators,
-    );
+
+    let periodic_columns = evaluator.build_periodic_values::<E, E>();
+    let periodic_columns_evaluations =
+        evaluate_periodic_columns_at(periodic_columns, &proof.0.openings_claim.eval_point);
+
+    let mut query = proof.0.openings_claim.openings.clone();
+    query.extend_from_slice(&periodic_columns_evaluations);
```

Comment:
Here we are assuming that periodic column values are always at the end of the query, right? But when we define oracles, I don't think this is enforced. Could this cause issues?

Reply:
Indeed we are, and this will break things if that is not the case. Should we add checks and panic throughout to enforce this?

Reply:
If we always expect it to be at the end of the `query` buffer, can't we simply not pass it to `build_query()`, and append it ourselves?

IIUC, there is no other correct way to do it, so there's no point in asking (and hoping for) the user to do it.

Reply:
Agreed. I think we should try to derive oracles for periodic columns so that the user does not need to make sure that oracles are ordered correctly.

Also, maybe we define query something like this:
```rust
pub struct Query<E: FieldElement> {
    pub committed_values: Vec<E>,
    pub periodic_value: Vec<E>,
}
```
This way, there is no ambiguity about where the periodic values should go (though, it does require 2 vectors instead of 1).

Reply:
> If we always expect it to be at the end of the `query` buffer, can't we simply not pass it to `build_query()`, and append it ourselves?
> 

I am not sure how `build_query()` fits into the code snippet above. `build_query()` is specifically for building from the main trace segment.

Reply:
> Agreed. I think we should try to derive oracles for periodic columns so that the user does not need to make sure that oracles are ordered correctly.
> 

I am not sure I understand fully but the way I see solving the issue raised is by separating the the oracles into committed and periodic. 
My assumption was that, if we are going to test the consistency between `get_oracles()` and `build_query()` anyway then we might as well impose that periodic columns are at the end. The test will check for this any and imposing that periodic columns are towards the end is implicit in requiring consistency between `get_oracles()` and `build_query()`.

Reply:
> I am not sure how build_query() fits into the code snippet above. build_query() is specifically for building from the main trace segment.

We are currently making an assumption about the `query` argument to `evaluate_query()` - namely that the periodic columns need to be appended at the end of the buffer. Hence, this assumption needs to be upheld by `build_query()` as well.

My point was: why give the user the possibly to screw it up and place the periodic columns somewhere other than at the end of the `query` buffer? We could instead change `build_query()` to be

```rs
    fn build_query<E>(&self, frame: &EvaluationFrame<E>, query: &mut [E])
    where
        E: FieldElement<BaseField = Self::BaseField>,
    { ... }
```

and append periodic values manually after each call to `build_query()`, for example [here](https://github.com/al-kindi-0/winterfell/blob/873345af6777668f88e245bcabcb995ccc28585c/prover/src/logup_gkr/mod.rs#L125-L126):

```rs
evaluator.build_query(&main_frame, &mut query);
query.extend(periodic_values_row);
```

i.e. we are enforcing that assumption ourselves, and not leaving the user any possibility for mistake.

IIUC, @irakliyk built on this and suggested we make the structure explicit by representing the periodic values separately (instead of appended at the end up the buffer). So the previous example would look like

```rs
evaluator.build_query(&main_frame, &mut query);
let query_full = Query::new(query, periodic_values_row);
```

Does that make sense? Maybe I'm missing something too.

Reply:
Yep, it does make sense. One thing that is different with respect to the new proposal is with respect to new allocations. In the current way, we just write to a buffer but with the current proposal I am not sure if we are incurring some kind of overhead when creating `Query`.

Reply:
I think `Query::new()` should be fine here - i.e., it would not allocate anything as it would leave on the stack.